### DISCUSSION TITLE
Self-hosted Supabase does not take into account changed base path, always serves on root path

### DISCUSSION THREAD

**Discussion by mariojose96** [supabase/supabase#33178] [Troubleshooting]
# Bug report

> * [x]  I confirm this is a bug with Supabase, not with my own application. 
> * [x]  I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

Supabase does not take into account a changed base path when changing SITE_URL, SUPABASE_PUBLIC_URL and API_EXTERNAL_URL. 

Trying to serve a self-hosted supabase instance via docker on a base path like http://localhost:8000/supabase will not work as the studio and api calls will always be served on http://localhost:8000.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Get latest supabase docker images following official supabase docs:
    ##### Get the code
    `git clone --depth 1 https://github.com/supabase/supabase`
    
    ##### Go to the docker folder
    `cd supabase/docker`
    
    ##### Copy the fake env vars
    `cp .env.example .env`
    
    ##### Pull the latest images
    `docker compose pull`
2. Edit .env and change `SUPABASE_PUBLIC_URL` and  `API_EXTERNAL_URL` to `http://localhost:8000/supabase` and `SITE_URL` to `http://localhost:3000/supabase`
3. Startup docker compose
    `docker compose up -d`
4. Go to `http://localhost:8000/supabase` and authenticate with default creds from supabase docs. 
5. See 404 Error on `http://localhost:8000/supabase`
6. Click on Head Back Button and see working Supabase Studio Dashboard on `http://localhost:8000` instead of `http://localhost:8000/supabase`

## Expected behavior

Supabase Studio should be served on `http://localhost:8000/supabase` when changing the `SITE_URL`  and `SUPABASE_PUBLIC_URL` to `http://localhost:8000/supabase`.

API calls should be called from `http://localhost:8000/supabase/api` instead of `http://localhost:8000/api`when changing the `API_EXTERNAL_URL` to `http://localhost:8000/supabase`.

## Screenshots

![404 on localhost/supabase](https://i.imgur.com/TFQrDbk.png)

![Working on root path localhost](https://i.imgur.com/3Qhx8Sf.png)

## System information

- OS: macOS Sonoma 14.6.1
- Browser: Tested on Safari, Chrome, Firefox

## Additional context

I tested this on my MacBook locally and also on a Ubuntu vps over a subdomain for hours.
In both cases it did not work with custom base paths, but on the root path.

Self-hosted Supabase works fine in general, but it is often the case that an instance should run on a specific path rather than the root path, as Supabase Cloud Studio does from supabase.com/dashboard.

**Comment by TimberBro** [comment]
Same problem on self-hosted coolify with dashboard behind kong.

**Comment by J-ARTS** [comment]
This issue is coming from the application itself, you guys have to fork the project and adjust the base path in the next.config.js and rebuild it.

**Comment by mariojose96** [comment]
You're right @J-ARTS, I almost forgot that I even posted this issue, but I've solved it in the meantime.
For anyone who has the same problem, here are the steps to temporarily fix it:

1. Git-clone the supabase repo if you haven't already done so
2. Inside the repo go to apps/studio/lib/constants and change this code:
`export const API_URL = IS_PLATFORM ? process.env.NEXT_PUBLIC_API_URL : '/api'` 
into this `export const API_URL =  process.env.NEXT_PUBLIC_API_URL`
3. Go to apps/studio/.env and define SUPABASE_PUBLIC_URL:  `SUPABASE_PUBLIC_URL=http://localhost:8000/supabase`.
This runs supabase from "/supabase", but you can change it to whatever you need it.
5. Then define NEXT_PUBLIC_API_URL for example to `NEXT_PUBLIC_API_URL=$SUPABASE_PUBLIC_URL/api`.
6. Build a local supabase studio docker image. From the root of the cloned repo run:
 `docker build . -f apps/studio/Dockerfile --target production -t studio:latest`
7. Go to docker/docker-compose.yml and add studio instead of prebuilt image:
```
studio:
    container_name: supabase-studio
    image: studio #supabase/studio:20240729-ce42139
```
8. Run `docker compose up -d ` from the docker folder

Et voilà 👍🏼

<img width="1896" alt="Supabase Self-Hosted Custom Basepath" src="https://github.com/user-attachments/assets/7b8df76d-ff57-46e8-a49a-1ff985fd7f0a" />

**Comment by Hallidayo** [comment]
Hi everyone, due to inactivity on this issue I've moved the issue over to discussions/enhancements.

**Comment by stefanadelbert** [comment]
I would also like to adjust the base path for Studio so that I can stick it behind a reverse proxy and use path-based routing to get to it.

**Comment by mhadeler** [comment]
Any update on this? The workaround posted by @mariojose96 doesn't seem to work anymore. Would really love to get this working.

**Comment by Alexei-Barnes** [comment]
This is impacting me as well. At this point, the application fails to load because the NextJS bundle contains a compiled base path of "", with no way to override it without source changes. If you fix that, the browser will try to talk to the api on `/api`, which doesn't take into account the base path. The workaround above no longer fixes this, because that code doesn't build. I think because the URL isn't valid server-side.

Is there a desire for a PR here?
I might have some free time to try and fix this.

Can it be converted back into an issue?
This seems like a bug to me.
